system/lib/wasmfs/syscalls.cpp

```diff
@@ -513,4 +520,85 @@ __wasi_errno_t __wasi_fd_fdstat_get(__wasi_fd_t fd, __wasi_fdstat_t* stat) {
   }
   return __WASI_ERRNO_SUCCESS;
 }
+
+// This enum specifies whether rmdir or unlink is being performed.
+enum class UnlinkMode { Rmdir, Unlink };
+
+static long __unlink(char* path, UnlinkMode unlinkMode) {
+  auto pathParts = splitPath(path);
+
+  // Root (/) directory cannot be removed.
+  if (pathParts.size() == 1 && pathParts[0] == "/") {
+    return -EBUSY;
+  }
+
+  if (pathParts.empty() || pathParts.back() == ".") {
+    return -EINVAL;
+  }
+
+  auto base = pathParts.back();
+
+  long err;
+  auto parentDir = getDir(pathParts.begin(), pathParts.end() - 1, err);
+
+  // Parent node doesn't exist.
+  if (!parentDir) {
+    return err;
+  }
+
+  // Hold the locked parent for the duration of the operation.
+  auto lockedParentDir = parentDir->locked();
+
+  auto curr = lockedParentDir.getEntry(base);
+
+  if (!curr) {
+    return -ENOENT;
+  }
+
+  // rmdir checks if the target is a directory and if the directory is empty.
+  auto targetDir = curr->dynCast<Directory>();
+  if (unlinkMode == UnlinkMode::Rmdir) {
+
+    if (!targetDir) {
+      return -ENOTDIR;
+    }
+
+    // The root directory or current work directory cannot be removed.
+    if (targetDir == wasmFS.getCWD()) {
+      return -EBUSY;
+    }
+
+    // A directory can only be removed if it has zero entries.
+    if (targetDir->locked().getNumEntries() > 0) {
+      return -ENOTEMPTY;
+    }
+  } else {
+    if (targetDir) {
+      return -EISDIR;
+    }
+  }
+
+  // Cannot unlink if the parent dir doesn't have write permissions.
+  if (!(lockedParentDir.mode() & 0222)) {
```

Comment:
Are there macros we can use instead of literal `0222`? Also, it might be reasonable to leave checking the permissions as a TODO. IIRC, we don't currently check them anywhere else.

Reply:
I think we can use a macro here. I wanted to include the checking permissions to match the existing test behaviour.